File: 48-rotate-image/48-rotate-image.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n):
            for j in range(i,n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        for j in range(n//2):
            for i in range(n):
                matrix[i][j], matrix[i][n-1-j] = matrix[i][n-1-j], matrix[i][j]
            
        
                
        

File: 48-rotate-image/48-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        //transpose first
        int nrow = matrix.length, ncol = matrix[0].length;
        
        for(int row = 0; row < nrow; row++)
            for(int col = 0; col < ncol; col++){
                if (row <= col) continue;
                matrix[row][col] = matrix[row][col] ^ matrix[col][row];
                matrix[col][row] = matrix[row][col] ^ matrix[col][row];
                matrix[row][col] = matrix[row][col] ^ matrix[col][row];
                
            }

        //exchange values of i and ncol - i
        for (int newCol = 0; newCol < ncol/2; newCol++)
            for(int newRow = 0; newRow < nrow; newRow++){
                int other_newCol = ncol - 1 - newCol;
                int temp = matrix[newRow][other_newCol];
                matrix[newRow][other_newCol] = matrix[newRow][newCol];
                matrix[newRow][newCol] = temp;
            }
        
        
        
    }
}